File: 24-swap-nodes-in-pairs/24-swap-nodes-in-pairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
    if(!head || !head.next)
        return head;
    let curr = head,nextNode = head.next;
    let next_pair = swapPairs(curr.next.next);
    nextNode.next=curr;
    curr.next=next_pair;
    return nextNode;
};

File: 24-swap-nodes-in-pairs/24-swap-nodes-in-pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        /* NAIVE APPROACH */
        
//         ListNode * p = head,*q; // pointers for iteration
//         ListNode *dummy = new ListNode(0);  // dummy node which help to swap nodes and change the value of head
// //         looping till we get null || till we get last odd node
//         while(p!=NULL && p->next!=NULL)
//         {
            
//             q = p->next;    
//             if(dummy->next == NULL) head = p->next;
//             dummy -> next = p -> next;
            
//             p->next=q->next;
//             q->next=p;
//             dummy = p;
//             p=p->next;
            
//         }
//         return head;
        
        /* RECURSIVE */
        if(head==NULL || head->next ==NULL)
            return head;
    
            ListNode *curr = head, *nextNode=head->next;
            ListNode *nextele = swapPairs(curr->next->next);
            nextNode->next=curr;
            curr->next=nextele;
    
            return nextNode;
    }
};